### DISCUSSION TITLE
Messaging contracts - is pact.v3 still the only way?

### DISCUSSION THREAD

**Discussion by Kamerrie** [pact-foundation/pact-python#762] [Q&A] [answered]
As title^

I apologise if this is considered bad GitHub/OSS etiquette - I'm new to the field. I'm basically trying to see if pact-python is valid for my use case.

Basically the company I work at is creating a new product, and I have been pushing extremely hard to use pact because our previous products used a combination of integrated tests and playwright, and I despised them as time went on for all the reasons that pact exists now. We will *mainly* have dotnet micro services, but there will be some python ones too, and we are going for event driven communication (think rabbitmq). The company would like to keep tests written in the same language as the services, which makes sense to me. After all, if a QE goes on leave a dev should still be capable of finishing a task.

With that in mind I've made a POC using the handy workshops and demos from pact net, and I can see a lot of info about messaging in pact python. I'm however seeing that in the readme messaging is only available in pact.v3, but in the tests folder we have 'from pact import messaging...' rather than from pact.v3, and there are examples of messaging both in and out of the V3 subfolder.

I'm trying to see if we can still work with pact-python in 2.2.1 state for now, if pact.v3 is stable yet (which I believe isn't until pact python v2.3 comes out? According to the tracking issue at least), if there are workarounds to test messaging in the current live version despite pact spec V2 not supporting it, or if we have to scrap pact for the python services until further notice.

**Comment by YOU54F** [comment, marked as answer]
Message-pact was introduced in the [pact-v3 specification](https://github.com/pact-foundation/pact-specification/tree/version-3?tab=readme-ov-file#introduces-messages-for-services-that-communicate-via-event-streams-and-message-queues)

it is available for use the original pact-python package, examples below

- https://github.com/pact-foundation/pact-python/blob/master/examples/tests/test_02_message_consumer.py
- https://github.com/pact-foundation/pact-python/blob/master/examples/tests/test_03_message_provider.py
- https://github.com/pactflow/example-consumer-python-sns

It uses [pact-message-ruby](https://github.com/pact-foundation/pact-message-ruby) under the hood and serialises to a 3.0.0 pact spec contract (async messages was only introduced in the pact v3 spec)

the `pact.v3` imports you are referring to, is a rework of pact-python to leverage the pact_ffi core, and async messaging has began to be introduced into pact-python, see blog post [here](https://pact-foundation.github.io/pact-python/blog/2024/07/26/asynchronous-message-support/)

the `pact.v3` import will ultimately allow you to create pact contracts which adhere to version 2, 3 or 4 of the pact-specification. There may just be some confusion between the version numbers of the pact-specification, and that of pact-python

**Reply by Kamerrie** [reply]
Thank you, this does exactly answer my question.

**Reply by YOU54F** [reply]
My pleasure, good luck on your pact journey, and be sure to reach out again if you are in need

**Comment by JP-Ellis** [comment]
Hi!

Please don't feel bad about asking the question. The GitHub discussion is certainly the right spot for questions such as yours.

As mentioned by @YOU54F, the main `pact` module _does_ support messages in the examples above, but does so in a way which we want to deprecate once Pact Python V3 is released.

I would certainly recommend you take a look at using the `pact.v3` module. @valkolovos, an amazing contributor to Pact python, has been doing a lot of implementation work to get asynchronous message interaction working, to the point where it passes Pact's compatibility suite for asynchronous messages.

The main ways I see `pact.v3` changing, in as far as asynchronous message go, is as follows:

- Not requiring the end-user to stand up a dummy HTTP server. Since we abstract away the specifics of the messaging system, we still need _some_ way to communicate, and it is simplest to use a dummy server. The long term goal here is that this HTTP server should be entirely hidden to the end user.

- Currently, the definitions have to be either strings or byte arrays, and support for matchers is done 'by hand' (that is, you write the JSON syntax of the matchers yourself, and serialize that to a JSON string). Work has started on that.

Neither of these should be breaking changes once implemented, but they do unfortunately make the experience right now lower than the standard I would like.

I believe @valkolovos is actually interested in this for his own work, and I am curious whether adoption of `pact.v3` has started. He might be able to speak about his experience.

**Reply by Kamerrie** [reply]
This is cool to understand! I might bring this up in the next auto guild and see if it's of interest for the others. In the end the current way is indeed being deprecated eventually so we'll have to move if we start with it. Thank you so much.
